src/wasm_opt.rs

```diff
@@ -0,0 +1,82 @@
+use anyhow::{ensure, Context, Result};
+use std::{path::PathBuf, process};
+
+#[allow(unreachable_code)]
+pub fn run(
+    binary: Vec<u8>,
+    shrink_level: u32,
+    optimization_level: u32,
+    debug_info: bool,
+) -> Result<Vec<u8>> {
+    return {
+        let wasm_opt = download_wasm_opt()?;
+
+        let mut command = process::Command::new(&wasm_opt);
+        command
+            .stderr(process::Stdio::inherit())
+            .args(&["-o", "-", "-O"])
+            .args(&["-ol", &optimization_level.to_string()])
+            .args(&["-s", &shrink_level.to_string()]);
+
+        if debug_info {
+            command.arg("-g");
+        }
+
+        #[cfg(target_os = "macos")]
+        {
+            command.env("DYLD_LIBRARY_PATH", wasm_opt.parent().unwrap());
+        }
+
+        #[cfg(windows)]
+        let delete_guard = {
+            use std::io::Write;
+
+            let tmp = tempfile::NamedTempFile::new()?;
+            tmp.as_file().write_all(&binary)?;
+            command.arg(tmp.path());
+            tmp
+        };
+
+        #[cfg(unix)]
+        {
+            use std::io::{Seek, SeekFrom, Write};
+
+            let mut file = tempfile::tempfile()?;
+            file.write_all(&binary)?;
+            file.seek(SeekFrom::Start(0))?;
+            command.stdin(file);
+        }
+
+        let output = command.output()?;
+        ensure!(output.status.success(), "command `wasm-opt` failed");
+
+        Ok(output.stdout)
+    };
+
+    log::warn!("No optimization has been done on the WASM");
+    Ok(binary)
+}
+
+fn download_wasm_opt() -> Result<PathBuf> {
```

Comment:
The whole thing should actually be in a lazy_static so we don't try to re-download if we optimize multiple wasm